File: flowcatalyst-go/internal/outbox/api_client.go
package outbox

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"time"
)

// APIClient sends batches of outbox items to the FlowCatalyst API
type APIClient struct {
	baseURL    string
	authToken  string
	httpClient *http.Client
}

// APIClientConfig holds configuration for the API client
type APIClientConfig struct {
	// BaseURL is the FlowCatalyst API base URL (required)
	BaseURL string

	// AuthToken is the optional Bearer token for authentication
	AuthToken string

	// ConnectionTimeout is the connection timeout
	ConnectionTimeout time.Duration

	// RequestTimeout is the request timeout
	RequestTimeout time.Duration
}

// DefaultAPIClientConfig returns sensible defaults
func DefaultAPIClientConfig() *APIClientConfig {
	return &APIClientConfig{
		ConnectionTimeout: 10 * time.Second,
		RequestTimeout:    30 * time.Second,
	}
}

// NewAPIClient creates a new API client
func NewAPIClient(config *APIClientConfig) *APIClient {
	if config == nil {
		config = DefaultAPIClientConfig()
	}

	return &APIClient{
		baseURL:   config.BaseURL,
		authToken: config.AuthToken,
		httpClient: &http.Client{
			Timeout: config.RequestTimeout,
		},
	}
}

// SendEventBatch sends a batch of events to the API
// POST /api/events/batch
func (c *APIClient) SendEventBatch(ctx context.Context, items []*OutboxItem) (*BatchResult, error) {
	return c.sendBatch(ctx, "/api/events/batch", items)
}

// SendDispatchJobBatch sends a batch of dispatch jobs to the API
// POST /api/dispatch/jobs/batch
func (c *APIClient) SendDispatchJobBatch(ctx context.Context, items []*OutboxItem) (*BatchResult, error) {
	return c.sendBatch(ctx, "/api/dispatch/jobs/batch", items)
}

// sendBatch sends a batch of items to the specified endpoint
func (c *APIClient) sendBatch(ctx context.Context, endpoint string, items []*OutboxItem) (*BatchResult, error) {
	if len(items) == 0 {
		return &BatchResult{}, nil
	}

	// Parse payloads into JSON array
	payloads := make([]json.RawMessage, len(items))
	for i, item := range items {
		payloads[i] = json.RawMessage(item.Payload)
	}

	body, err := json.Marshal(payloads)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal batch: %w", err)
	}

	url := c.baseURL + endpoint
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	if c.authToken != "" {
		req.Header.Set("Authorization", "Bearer "+c.authToken)
	}

	slog.Debug("Sending batch to API",
		"endpoint", endpoint,
		"batchSize", len(items))

	resp, err := c.httpClient.Do(req)
	if err != nil {
		result := NewBatchResult()
		result.Error = err
		// Mark all as failed with internal error status
		for _, item := range items {
			result.FailedItems[item.ID] = StatusInternalError
		}
		return result, err
	}
	defer resp.Body.Close()

	// Read response body
	respBody, _ := io.ReadAll(io.LimitReader(resp.Body, 64*1024))

	if resp.StatusCode >= 400 {
		err := fmt.Errorf("API returned status %d: %s", resp.StatusCode, string(respBody))
		slog.Error("API batch request failed",
			"statusCode", resp.StatusCode,
			"endpoint", endpoint,
			"response", string(respBody))
		result := NewBatchResult()
		result.Error = err
		// Determine status based on HTTP code
		status := StatusFromHTTPCode(resp.StatusCode)
		for _, item := range items {
			result.FailedItems[item.ID] = status
		}
		return result, err
	}

	slog.Debug("Batch sent successfully",
		"endpoint", endpoint,
		"batchSize", len(items),
		"statusCode", resp.StatusCode)

	result := NewBatchResult()
	result.SuccessIDs = extractIDs(items)
	return result, nil
}

// extractIDs extracts IDs from a slice of items
func extractIDs(items []*OutboxItem) []string {
	ids := make([]string, len(items))
	for i, item := range items {
		ids[i] = item.ID
	}
	return ids
}

File: flowcatalyst-typescript/apps/flowcatalyst/src/outbox-processor/api-client.ts
/**
 * FlowCatalyst API Client
 *
 * HTTP client for the platform's batch API endpoints.
 * Sends batches of outbox items and maps HTTP responses to OutboxStatus.
 */

import { OutboxStatus, httpStatusToOutboxStatus, type OutboxItem, type OutboxItemType } from './model.js';
import type { OutboxProcessorConfig } from './env.js';
import type { Logger } from 'pino';

/**
 * Result of a batch API call.
 */
export interface BatchResult {
  /** Whether all items succeeded. */
  readonly allSuccess: boolean;
  /** Map of failed item IDs to their OutboxStatus. */
  readonly failedItems: Map<string, OutboxStatus>;
  /** Error message if the entire batch failed. */
  readonly errorMessage: string | null;
}

function allSuccess(): BatchResult {
  return { allSuccess: true, failedItems: new Map(), errorMessage: null };
}

function allFailed(ids: string[], status: OutboxStatus, errorMessage: string): BatchResult {
  const failedItems = new Map<string, OutboxStatus>();
  for (const id of ids) {
    failedItems.set(id, status);
  }
  return { allSuccess: false, failedItems, errorMessage };
}

export interface ApiClient {
  createEventsBatch(items: OutboxItem[]): Promise<BatchResult>;
  createDispatchJobsBatch(items: OutboxItem[]): Promise<BatchResult>;
  createAuditLogsBatch(items: OutboxItem[]): Promise<BatchResult>;
}

export function createApiClient(config: OutboxProcessorConfig, logger: Logger): ApiClient {
  const { apiBaseUrl, apiToken } = config;

  async function post(
    path: string,
    items: OutboxItem[],
  ): Promise<BatchResult> {
    const ids = items.map((item) => String(item.id));
    const payloads = items.map((item) => {
      try {
        return JSON.parse(item.payload);
      } catch {
        throw new Error(`Invalid JSON payload for item ${item.id}`);
      }
    });

    const url = `${apiBaseUrl}${path}`;
    const body = JSON.stringify({ items: payloads });

    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };
    if (apiToken) {
      headers['Authorization'] = `Bearer ${apiToken}`;
    }

    try {
      const controller = new AbortController();
      const timeout = setTimeout(() => controller.abort(), 30_000);

      const response = await fetch(url, {
        method: 'POST',
        headers,
        body,
        signal: controller.signal,
      });

      clearTimeout(timeout);

      const statusCode = response.status;

      if (statusCode >= 200 && statusCode < 300) {
        logger.debug({ path, statusCode }, 'Batch API call succeeded');
        return allSuccess();
      }

      const responseBody = await response.text().catch(() => '');
      const outboxStatus = httpStatusToOutboxStatus(statusCode);
      const errorMessage = `API error: ${statusCode} - ${responseBody}`;
      logger.error({ path, statusCode, responseBody }, 'Batch API call failed');

      return allFailed(ids, outboxStatus, errorMessage);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);

      if (error instanceof Error && error.name === 'AbortError') {
        logger.error({ path }, 'Batch API call timed out');
        return allFailed(ids, OutboxStatus.GATEWAY_ERROR, `Request timeout: ${message}`);
      }

      logger.error({ path, error: message }, 'Batch API call failed with network error');
      return allFailed(ids, OutboxStatus.GATEWAY_ERROR, `IO error: ${message}`);
    }
  }

  return {
    createEventsBatch(items) {
      return post('/api/events/batch', items);
    },
    createDispatchJobsBatch(items) {
      return post('/api/dispatch/jobs/batch', items);
    },
    createAuditLogsBatch(items) {
      return post('/api/audit-logs/batch', items);
    },
  };
}
